### PR TITLE
Feat/Support license type selection when purchasing privileges

### LINKED ISSUE #572: Multiple licenses - privilege purchase validation
We need to update the privilege purchase API to support license type selection.

Tasks:
- Update privilege purchase API schema to include license type selection
- Implement backend validation for license type selection, based on active home state licenses
- Store selected license type with privilege record

### PR DESCRIPTION
**Pull request by landonshumway-ia**
As part of supporting multiple license types in the same home state, this adds a `licenseType` field to the request body of the privilege purchase endpoint. The endpoint now validates that the specified license type matches one of the provider's home state licenses before proceeding with the privilege purchase.

### Requirements List
- This is a backwards incompatible change, when this is merged the UI will not be able to purchase privileges until [this PR for the corresponding front end change is merged. ](https://github.com/csg-org/CompactConnect/pull/627)

### Description List
- Add required `licenseType` field to POST privilege purchase request body
- Validation license type matches one of provider's home state licenses

### Testing List
- updated api schema snapshots
- updated smoke tests and verified change in sandbox env
- functional/unit test coverage added for validation logic
- Code review

Closes #572

### CODE REVIEW

File: backend/compact-connect/lambdas/python/purchases/handlers/privileges.py [hunk 1]
```diff
@@ -211,21 +212,29 @@ def post_purchase_privileges(event: dict, context: LambdaContext):  # noqa: ARG0
...
 
-    home_state_selection = ProviderRecordUtility.get_provider_home_state_selection(user_provider_data['items'])
-    if home_state_selection is None:
-        raise CCInternalException('No home state selection found for this user')
-
-    home_state_license_record = ProviderRecordUtility.find_best_license(
-        license_records=license_records,
-        home_jurisdiction=home_state_selection,
+    # we now validate that the license type matches one of the license types from the home state license records
+    matching_license_record = next(
```
Comment (jusdino) [inline backend/compact-connect/lambdas/python/purchases/handlers/privileges.py:225]: You can combine this license record filtering step with your `ProviderRecordUtility.get_records_of_type()` call by leveraging its `_filter` kwarg.

File: backend/compact-connect/stacks/api_stack/v1_api/api_model.py [hunk 2]
```diff
@@ -492,8 +492,12 @@ def post_purchase_privileges_request_model(self) -> Model:
             description='Post purchase privileges request model',
             schema=JsonSchema(
                 type=JsonSchemaType.OBJECT,
-                required=['selectedJurisdictions', 'orderInformation', 'attestations'],
+                required=['licenseType', 'selectedJurisdictions', 'orderInformation', 'attestations'],
                 properties={
+                    'licenseType': JsonSchema(
```
Comment (jusdino) [inline backend/compact-connect/stacks/api_stack/v1_api/api_model.py:497]: Let's set this as an enum of license types.

**Review by jusdino — CHANGES_REQUESTED** [review]
Looks good! I'd just suggest a couple of minor tweaks.

**Review by jusdino — APPROVED** [review]
Looks good!

**Comment by landonshumway-ia** [comment]
@jlkravitz This is ready for your review. Thanks!

File: backend/compact-connect/lambdas/python/common/tests/function/test_data_client.py [hunk 3]
```diff
@@ -221,7 +221,13 @@ def test_data_client_created_privilege_record(self):
...
 
     @patch('cc_common.config._Config.current_standard_datetime', datetime.fromisoformat('2024-11-08T23:59:59+00:00'))
-    def test_data_client_updates_privilege_records(self):
+    def test_data_client_updates_privilege_records_for_specific_license_type(self):
+        """
+        In this test case, a user has two license types, audiologist and speech-language pathologist.
+        The user is purchasing a renewal for their audiologist privilege in ky, and purchasing a new
+        audiologist privilege in ne. The privilege for the speech-language pathologist license should not be
+        referenced nor updated in any way as part of this purchase.
+        """
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/common/tests/function/test_data_client.py:230]: Thanks for spelling this out- the edge cases are getting tricky!

File: backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py [hunk 4]
```diff
@@ -336,13 +338,54 @@ def test_purchase_privileges_invalid_if_existing_privilege_expiration_matches_li
...
 
+    @patch('handlers.privileges.PurchaseClient')
+    def test_purchase_privileges_valid_even_if_existing_privilege_for_another_license_type_has_same_expiration(
+        self, mock_purchase_client_constructor
+    ):
+        """
+        In this case, the user is attempting to purchase a privilege in kentucky for a new license type,
+        the expiration date for the other license type has not been updated since the last renewal, but the expiration
+        date for the selected license type has been updated since the last renewal. This purchase should be allowed.
+        """
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py:356]: A bit confused by this description/test, particularly the part that says, "the expiration date for the other license type has not been updated since the last renewal, but the expiration date for the selected license type has been updated since the last renewal."

Is this just testing that the expiration date for a privilege of a particular license type doesn't affect a provider's ability to purchase a privilege for _another_  license type (in the same jurisdiction), when it would otherwise if the purchase was for the same license type, since that privilege is active?

File: backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py [hunk 5]
```diff
@@ -608,6 +670,26 @@ def test_post_purchase_privileges_stores_attestations_in_privilege_record(self,
 
         self.assertEqual(generate_default_attestation_list(), privilege_record['attestations'])
 
+    @patch('handlers.privileges.PurchaseClient')
+    @patch('cc_common.config._Config.current_standard_datetime', datetime.fromisoformat('2024-11-08T23:59:59+00:00'))
+    def test_post_purchase_privileges_stores_license_type_in_privilege_record(self, mock_purchase_client_constructor):
+        """Test that attestations are stored in the privilege record."""
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py:676]: outdated comment?

File: backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py [hunk 6]
```diff
@@ -608,6 +670,26 @@ def test_post_purchase_privileges_stores_attestations_in_privilege_record(self,
...
+
+        self._when_purchase_client_successfully_processes_request(mock_purchase_client_constructor)
+
+        event = self._when_testing_provider_user_event_with_custom_claims(license_expiration_date='2050-01-01')
+        event['body'] = _generate_test_request_body()
+
+        resp = post_purchase_privileges(event, self.mock_context)
+        self.assertEqual(200, resp['statusCode'], resp['body'])
+
+        # check that the privilege record for ky was created with attestations
```
Comment (jlkravitz) [inline backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py:687]: outdated comment?

**Review by jlkravitz — COMMENTED** [review]
looks good! couple questions.

File: backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py (hunk 5 above)
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py:676]: yep, copy/paste. I'll update it

File: backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py (hunk 4 above)
Comment (landonshumway-ia) [reply backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py:356]: This is checking a rare edge case where a user has two licenses (we'll call them A and B) which happen to have the **exact** same expiration date (highly unlikely, but not impossible). The user also has a privilege for license A. 

Suppose the user wants to buy a privilege in the same jurisdiction for license B. Prior to this update, the endpoint would fail since it would see the expiration date of the privilege for license A and incorrectly return a 400, stating that a privilege already exists for that jurisdiction with that expiration date.

I'll see if I can reword this comment to make it more clear

Comment (jlkravitz) [reply backend/compact-connect/lambdas/python/purchases/tests/function/test_handlers/test_purchase_privileges.py]: Ah, got it. The new comment is clear and makes a lot of sense. Thanks for updating.

**Review by jlkravitz — APPROVED** [review]
@isabeleliassen This is good to merge.

**Comment by landonshumway-ia** [comment]
@jlkravitz @isabeleliassen - @ChiefStief just identified a potential issue during testing this on the frontend. I will need to add the fix for it before this should be merged

**Review by jlkravitz — APPROVED** [review]
@landonshumway-ia Looks good!

**Review by isabeleliassen — APPROVED** [review]
